### PR TITLE
feat(dyn-abi): `DynSolType::coerce_str`

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

Port of [`ethabi::Tokenizer`](https://docs.rs/ethabi/latest/ethabi/token/trait.Tokenizer.html), with extra features

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [x] Breaking changes

### CODE REVIEW

File: crates/dyn-abi/src/coerce.rs [hunk 1]
```diff
@@ -0,0 +1,861 @@
...
+    let Some(delim) = input.chars().next() else {
+        return Ok("")
+    };
+    let has_delim = matches!(delim, '"' | '\'');
+    if has_delim {
+        *input = &input[1..];
+    }
+
+    // TODO: escapes?
+    let min = if has_delim { 0 } else { 1 };
```
Comment (mattsse) [inline crates/dyn-abi/src/coerce.rs:328]: hmm, good question, maybe this could happen with same escaped json inputs, but I think we can ignore that

**Review by mattsse — APPROVED** [review]
amazing work!

it's been a while since I read some nom code but I still love how readable this is.

test coverage looks decent.

this will simplify so much

File: crates/dyn-abi/src/coerce.rs [hunk 2]
```diff
@@ -0,0 +1,861 @@
...
+fn uint_units(input: &mut &str) -> PResult<usize> {
+    alpha0(input).and_then(|s| {
+        Ok(match s {
+            "ether" => 18,
+            "gwei" | "nano" | "nanoether" => 9,
+            "" | "wei" => 0,
+            _ => return Err(ErrMode::from_error_kind(input, ErrorKind::Fail)),
+        })
+    })
+}
```
Comment (Evalir) [inline crates/dyn-abi/src/coerce.rs:263]: This is super handy, lets us replace a few utilities we have in foundry

wdyt about adding more units?

File: crates/dyn-abi/src/coerce.rs [hunk 3]
```diff
@@ -0,0 +1,861 @@
...
+
+        if cfg!(feature = "std") {
+            assert_eq!(
+                DynSolType::Uint(256).coerce_str("0.01 ether").unwrap(),
+                DynSolValue::Uint(U256::from_str("10000000000000000").unwrap(), 256)
+            );
+
+            assert_eq!(
+                DynSolType::Uint(256)
+                    .coerce_str("0.000000000000000001ether")
```
Comment (Evalir) [inline crates/dyn-abi/src/coerce.rs:606]: nice

**Review by Evalir — APPROVED** [review]
Super readable! This is gonna be super handy—love it

File: crates/dyn-abi/src/coerce.rs (hunk 2 above)
Comment (DaniPopes) [reply crates/dyn-abi/src/coerce.rs:263]: Do you have anything in mind? Solidity only supports `wei`, `gwei`, and `ether`

**Review by prestwich — COMMENTED** [review]
do we have documentation of all the formats supported?

**Comment by DaniPopes** [comment]
Added in 0e974a13502f264c4ded94d36395123b3fe628e0 @prestwich

**Comment by prestwich** [comment]
perfection

**Pull request merged**